oshdb-api/src/main/java/org/heigit/ohsome/oshdb/api/mapreducer/MapReducer.java

```diff
@@ -866,19 +597,42 @@ public MapReducer<X> filter(String f) {
    */
   @Contract(pure = true)
   public MapReducer<List<X>> groupByEntity() throws UnsupportedOperationException {
-    if (!this.mappers.isEmpty()) {
-      throw new UnsupportedOperationException(
-          "groupByEntity() must be called before any `map` or `flatMap` "
-              + "transformation functions have been set");
-    }
     if (this.grouping != Grouping.NONE) {
       throw new UnsupportedOperationException("A grouping is already active on this MapReducer");
     }
-    MapReducer<X> ret = this.copy();
-    ret.grouping = Grouping.BY_ID;
-    @SuppressWarnings("unchecked") // now in the reduce step the backend will return a list of items
-    MapReducer<List<X>> result = (MapReducer<List<X>>) ret;
-    return result;
+    if (!this.mappers.isEmpty()) {
+      // for convenience, we allow one to set this function even after some map functions were set.
+      // if some map / flatMap functions were already set:
+      // "rewind" them first, apply the grouping and then re-apply the map/flatMap functions
+      // accordingly
+      MapReducer<X> ret = this.copy();
+      List<MapFunction> mappers = new LinkedList<>(ret.mappers);
```

Comment:
```suggestion
      List<MapFunction> mappers = new ArrayList<>(ret.mappers);
```

Reply:
Also rename `mappers` to something else as this variable hides the class variable `mappers` and this is reported as a code smell. The same is true in `aggregateByGeometry`.

Reply:
Found it in the aforementioned methods and `aggregateByTimestamp`, see 209ae2c3.

Reply:
fixed in 40dba317 by simplifying the implementation of the "automatic" aggregation methods. As as sideeffect, now all(?) aggregation combinations are allowed and possible regardless of their order, for example: `mapRed.aggregateByGeometry(…).aggregateByTimestamp()` or `mapRed.map(…).aggregateByTimestamp().aggregateByGeometry(…)`